### ISSUE TITLE
Issue with sqlcmd connection

### ISSUE THREAD

**Issue by nkotha88** [MicrosoftDocs/sql-docs#9240]
$ sqlcmd -S localhost -U sa
Password: 
Sqlcmd: Error: Microsoft ODBC Driver 18 for SQL Server : SSL Provider: [error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed:self signed certificate].
Sqlcmd: Error: Microsoft ODBC Driver 18 for SQL Server : Client unable to establish connection. For solutions related to encryption errors, see https://go.microsoft.com/fwlink/?linkid=2226722.

Have tried multiple options setting env variables, disabling encryption changes to MSSQL config file, disabling ODBC driver connection encryption. Nothing works. Some serious documentation issues with the images provided for MSSQL for Linux images listed in Azure portal. Please reach out to nkotha@microsoft.com

**Comment by rwestMSFT** [comment]
There is no issue with `sqlcmd`. As documented in multiple areas, TDE 8, which is the new version of the protocol underlying connections to SQL Server and Azure SQL Database / SQL Managed Instance, requires a trusted connection.

> Newer versions of sqlcmd are secure by default. For more information about connection encryption, see [sqlcmd utility](https://learn.microsoft.com/en-us/sql/tools/sqlcmd/sqlcmd-utility?view=sql-server-ver16) for Windows, and [Connecting with sqlcmd](https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/connecting-with-sqlcmd?view=sql-server-ver16) for Linux and macOS. If the connection doesn't succeed, you can add the -No option to sqlcmd to specify that encryption is optional, not mandatory.

**Comment by nkotha88** [comment]
Hi Randolph,

Thank you for reaching out.

I have just replied to the other thread below. ==> I do not think SQLCMD is linked in this doc? https://learn.microsoft.com/en-us/azure/azure-sql/virtual-machines/linux/sql-vm-create-portal-quickstart?view=azuresql-vm
but it is here  https://learn.microsoft.com/en-us/sql/linux/quickstart-install-connect-red-hat?view=sql-server-ver16&tabs=rhel8


https://learn.microsoft.com/en-us/azure/azure-sql/virtual-machines/linux/sql-vm-create-portal-quickstart?view=azuresql-vm is the doc which should be followed in case you are using Azure VM Image of SQL.

============================================== Email to other thread for more clarity =========================


Here are a few things that I as a new bee to SQL on Azure Linux have faced during VM image SQL Server 2022 on RHEL HA-8.6 deployment and then configuring it.
Coming from Oracle and using Linux with Oracle/PostgreSQL and some SQL on Windows experience I feel the following changes could help a customer starting out to explore SQL on Linux Journey.
Looking forward to hear anything I did not do as best practice.


  1.
Below command does not work as we can see, probably a typo?

https://learn.microsoft.com/en-us/sql/linux/quickstart-install-connect-red-hat?view=sql-server-ver16&tabs=rhel8


Install the SQL Server command-line tools

sudo yum install -y mssql-tools18 unixodbc-dev ==> This one needs to be corrected if not you will get the following error.

***@***.*** ~]$ sudo yum install -y mssql-tools18 unixodbc-dev
Repository packages-microsoft-com-prod is listed more than once in the configuration
Repository packages-microsoft-com-mssql-server-2022 is listed more than once in the configuration
Last metadata expiration check: 0:13:35 ago on Mon 14 Aug 2023 05:45:34 AM UTC.
No match for argument: unixodbc-dev
Error: Unable to find a match: unixodbc-dev

WA: Use below cmmands.

sudo yum install mssql-tools18 unixODBC-devel
or sudo ACCEPT_EULA=Y yum install -y mssql-tools18 unixODBC-devel



  1.  Let's say I followed the correct doc https://learn.microsoft.com/en-us/azure/azure-sql/virtual-machines/linux/sql-vm-create-portal-quickstart?view=azuresql&viewFallbackFrom=sql-server-ver1<https://learn.microsoft.com/en-us/azure/azure-sql/virtual-machines/linux/sql-vm-create-portal-quickstart?view=azuresql&viewFallbackFrom=sql-server-ver16>

after deployment of VM. SQLCMD -No option is still an issue, and you need to refer to https://learn.microsoft.com/en-us/sql/linux/quickstart-install-connect-red-hat?view=sql-server-ver16&tabs=rhel8
again, for references/troubleshooting. Since we do not have enough info on what is installed on the server and the best practices and references based on various versions of SQL and ODBC config.
For example sqlcmd -No -S localhost -U sa does not work if we do not reinstall ODBC with 18 version for SQL Server 2022 on RHEL HA-8.6 image.

 ***@***.*** ~]$
***@***.*** ~]$ sudo systemctl stop mssql-server
***@***.*** ~]$ sudo /opt/mssql/bin/mssql-conf set-sa-password
Enter the SQL Server system administrator password:
Confirm the SQL Server system administrator password:
Configuring SQL Server...

ForceFlush is enabled for this instance.
ForceFlush feature is enabled for log durability.
The system administrator password has been changed.
Please run 'sudo systemctl start mssql-server' to start SQL Server.
***@***.*** ~]$
***@***.*** ~]$
***@***.*** ~]$
***@***.*** ~]$ sudo systemctl start mssql-server
***@***.*** ~]$
***@***.*** ~]$
***@***.*** ~]$ echo 'export PATH="$PATH:/opt/mssql-tools/bin"' >> ~/.bash_profile
***@***.*** ~]$ echo 'export PATH="$PATH:/opt/mssql-tools/bin"' >> ~/.bashrc
***@***.*** ~]$ source ~/.bashrc
***@***.*** ~]$
***@***.*** ~]$ sudo firewall-cmd --zone=public --add-port=1433/tcp --permanent
success
***@***.*** ~]$ sudo firewall-cmd --reload
success
***@***.*** ~]$ sqlcmd -No -S localhost -U sa
Password:
Sqlcmd: Error: Microsoft ODBC Driver 17 for SQL Server : SSL Provider: [error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed:self signed certificate].
Sqlcmd: Error: Microsoft ODBC Driver 17 for SQL Server : Client unable to establish connection.
***@***.*** ~]$
***@***.*** ~]$
***@***.*** ~]$ sqlcmd -C -S localhost -U sa  ==> Again -C options works with this version of ODBC which appears to be 17 which comes default with the image.
Password:
1>
2>
3>

Work Around: for -No part of SQLCMD: Go to doc https://learn.microsoft.com/en-us/sql/linux/quickstart-install-connect-red-hat?view=sql-server-ver16&tabs=rhel8
and start at ODBC part to remove and readd properly with version 18.

sudo ACCEPT_EULA=Y yum install -y mssql-tools18 unixODBC-devel
sudo yum check-update
sudo yum update mssql-tools18

Now below works:

***@***.*** ~]$ echo 'export PATH="$PATH:/opt/mssql-tools18/bin"' >> ~/.bash_profile
***@***.*** ~]$
***@***.*** ~]$ echo 'export PATH="$PATH:/opt/mssql-tools18/bin"' >> ~/.bashrc
***@***.*** ~]$ source ~/.bashrc
***@***.*** ~]$
***@***.*** ~]$ sqlcmd -No -S localhost -U sa
Password:
1>
2>
3>



  1.  Highlight that the admin user created during the installation is only limited to Linux OS level super user or can be used as sudo. The need for sa user
right after logging into the new deployment. Any references to details of tools actually installed with the image like the binary's/log directory structures and tools etc.

            Even referencing some links to SQLCMD usage
https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/connecting-with-sqlcmd?view=sql-server-ver16
            in all the docs.



        4)  If you do not stumble on  https://learn.microsoft.com/en-us/azure/azure-sql/virtual-machines/linux/sql-vm-create-portal-quickstart?view=azuresql&viewFallbackFrom=sql-server-ver1<https://learn.microsoft.com/en-us/azure/azure-sql/virtual-machines/linux/sql-vm-create-portal-quickstart?view=azuresql&viewFallbackFrom=sql-server-ver16>
and configure using
https://learn.microsoft.com/en-us/sql/linux/quickstart-install-connect-red-hat?view=sql-server-ver16&tabs=rhel8 then you will find many options pretty confusing
and after some extensive time, you would look for other documents like the QuickStart guide.
sudo /opt/mssql/bin/mssql-conf setup ==> especially for this step as it differs a lot in both docs.


       5) We do not have any tutorial links/blogs/videos pointing out which docs to follow for SQL Server Image on Linux post deployment. All the content was generic to VM
  troubleshooting and Maintenace etc.

 ==> It was after I deployed it a few times troubleshooting I saw the plans section under the image creation tabs which had the following. This could be changed to refer to the post

  1.  deployment doc for better guidance.


[cid:a6309527-ad97-455e-b2d7-02ea65c71509]

[cid:4c8ce77b-f472-46a7-8987-0ef34143b353]


Thanks,
Nitish



________________________________

**Comment by nkotha88** [comment]
Yes, I’m an employee with the DP Ninja SQL engineering group of Microsoft.

I also need a few insights into what I said and how we can execute the procedures for example like sp_helpdb within Linux. Any doc covering this comparison with Windows SQL commands and features.

Regards,
Nitish
________________________________

**Comment by nkotha88** [comment]
Got it. Thank you. Will take it internally.

Regards,
Nitish
________________________________
